curated_transformers/models/scalar_weight.py

```diff
@@ -0,0 +1,147 @@
+            torch.tensor((1.0,), requires_grad=True)
+        )
+        self.dropout = torch.nn.Dropout(p=dropout_prob)
+
+    @property
+    def num_layers(self) -> int:
+        return self.layer_weights.size()[0]
+
+    @property
+    def dropout_prob(self) -> float:
+        return self.dropout.p
+
+    def forward(
+        self,
+        layer_outputs: List[Tensor],
+    ) -> Tensor:
+        """
+        Shapes:
+            layer_outputs - List[(batch, seq_len, layer_size)]
+
+        Returns a weighted tensor of the inputs with shape (batch, seq_len, layer_size).
+        """
+        if len(layer_outputs) != self.layer_weights.shape[0]:
+            raise ValueError(
+                f"Expected {self.layer_weights.shape[0]} layers, got {len(layer_outputs)} instead"
+            )
+
+        # (batch, seq_len, num_layers, layer_size)
+        stacked = torch.stack(layer_outputs, dim=2)
+
+        if self.training:
+            dropout_mask = torch.full_like(
+                self.layer_weights, 1.0 - self.dropout.p
+            ).bernoulli()
+            softmask_mask = (1.0 - dropout_mask) * -10000.0
+            layer_weights = self.layer_weights + softmask_mask
+        else:
+            layer_weights = self.layer_weights
+
+        # Convert the layer weights into a probability distribution and
+        # expand dimensions to get shape [1, 1, n_layers, 1].
+        layer_weights = (
+            layer_weights.softmax(dim=-1).unsqueeze(0).unsqueeze(0).unsqueeze(-1)
+        )
+        weighted_layers = stacked * layer_weights
+
+        return weighted_layers.sum(dim=-2, keepdim=False) * self.scale
+
+
+def build_scalar_weight_v1(
+    *,
+    num_layers: int,
+    dropout_prob: float = 0.1,
+    mixed_precision: bool = False,
+    grad_scaler_config: dict = SimpleFrozenDict(),
+) -> Model[List[Ragged], Ragged]:
+    if isinstance(grad_scaler_config, SimpleFrozenDict):
+        # Create a new, mutable dict instance.
+        grad_scaler_config = {}
+
+    if "enabled" not in grad_scaler_config:
+        grad_scaler_config["enabled"] = mixed_precision
+
+    # Increment number of layers by one to include the embedding layer.
+    scalar_weighting_layer = ScalarWeight(
+        num_layers=num_layers + 1, dropout_prob=dropout_prob
+    )
+
+    model = PyTorchWrapper_v2(
+        scalar_weighting_layer,
+        convert_inputs=_convert_inputs,
+        convert_outputs=_convert_outputs,
+        mixed_precision=mixed_precision,
+        grad_scaler=PyTorchGradScaler(**grad_scaler_config),
+    )
+
+    return model
+
+
+def _convert_inputs(model: Model, X: List[Ragged], is_train: bool = False):
+    ops = model.ops
+    layer_hidden_sizes = [x.data.shape[1] for x in X]
+    if not all_equal(layer_hidden_sizes):
+        raise ValueError(
+            f"Not all hidden sizes are equal in input passed to scalar weight"
+        )
+
+    Xt = []
+    for layer in X:
+        batch = ops.alloc3f(1, layer.data.shape[0], layer.data.shape[1])
```

Comment:
Is it necessary to have the initial singleton dimension? Maybe it's better to modify the PyTorch model to not require this dimension?

Another thing that I wondered about -- since we are copying here anyway, maybe we should allocate + fill a single tensor for all layers, so that the `stack` is not necessary in the PyTorch model?

Reply:
I originally had it implemented in such a way that we'd pass the entire document batch as its input. But it turned out to be unwieldy as the pooling didn't support batching. I'll stack the tensors without the batch dim during input conversion.